### ISSUE TITLE
In `loadShader()` casting between is an undefined behavior.

### ISSUE THREAD

**Issue by nikitablack** [SaschaWillems/Vulkan#592]
According to [strict aliasing rule](https://en.cppreference.com/w/cpp/language/reinterpret_cast) casting a pointer type to another pointer type with a more strict alignment is undefined behavior. So in [loadShader()](https://github.com/SaschaWillems/Vulkan/blob/master/base/VulkanTools.cpp) casting `char *shaderCode` to `moduleCreateInfo.pCode` (which is `uint32_t*`) is undefined behavior.

**Comment by polarina** [comment]
The strict aliasing rule has a special exception for `char` pointers. Char pointers may alias any other pointer (except restrict pointers).

**Comment by nikitablack** [comment]
That's true. But in the function the alias is a `uint32_t*` pointer, not `char`.

**Comment by polarina** [comment]
The original pointer is a `char` pointer and may alias. The cast is valid as long as the alignment requirements of `uint32_t` are satisfied, which they are since all allocations from `new` satisfy `alignof(max_align_t)`.

**Comment by nikitablack** [comment]
Why do you think that `char` pointer can be aliased? I'd be very grateful for the link to the standard or similar. I'm not a language lawyer and my reasonings are based exclusively on cppreference documentation, stackoverflow questions, and numerous articles.

For example, the aforementioned [Type aliasing](https://en.cppreference.com/w/cpp/language/reinterpret_cast) section enumerates all valid cases. And it doesn't mention I can cast from char. But I can do the opposite way (`AliasedType is std::byte, char, or unsigned char: this permits examination of the object representation of any object as an array of bytes`).

Additionally, [this](https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8#type-punning-arrays) very often quoted article has a similar to our discussion example where it shows that casting `char array` to an `int` violates strict aliasing.

**Comment by SaschaWillems** [comment]
Closing this due to inactivity. If the issue still persist, please open a new issue.
